java.lang.ClassNotFoundException on grpc shaded with guava-android in dependencies

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

I have project with grpc-netty-shaded lib that depends on guava 33.5.0-android which don't contain com.google.common.collect.SingletonImmutableBiMap. So it cause exception:
```
java.lang.RuntimeException: java.lang.ClassNotFoundException: com.google.common.collect.SingletonImmutableBiMap
	at org.apache.fory.serializer.collection.GuavaCollectionSerializers.loadClass(GuavaCollectionSerializers.java:713)
	at org.apache.fory.serializer.collection.GuavaCollectionSerializers.registerDefaultSerializers(GuavaCollectionSerializers.java:632)
	at org.apache.fory.resolver.ClassResolver.addDefaultSerializers(ClassResolver.java:384)
	at org.apache.fory.resolver.ClassResolver.initialize(ClassResolver.java:331)
	at org.apache.fory.Fory.<init>(Fory.java:155)
	at org.apache.fory.config.ForyBuilder.newFory(ForyBuilder.java:740)
	at org.apache.fory.config.ForyBuilder.lambda$factory$0(ForyBuilder.java:822)
	at org.apache.fory.pool.ThreadPoolFory.lambda$new$0(ThreadPoolFory.java:64)
	at org.apache.fory.pool.ThreadPoolFory.<init>(ThreadPoolFory.java:69)
	at org.apache.fory.config.ForyBuilder.buildThreadSafeFory(ForyBuilder.java:774)
---
Caused by: java.lang.ClassNotFoundException: com.google.common.collect.SingletonImmutableBiMap
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:580)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:490)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Class.java:547)
	at org.apache.fory.serializer.collection.GuavaCollectionSerializers.loadClass(GuavaCollectionSerializers.java:711)
	... 37 more
```


From dependency tree:
```
[INFO] +- io.grpc:grpc-netty-shaded:jar:1.79.0:runtime
[INFO] |  +- io.grpc:grpc-util:jar:1.79.0:runtime
[INFO] |  +- com.google.guava:guava:jar:33.5.0-android:compile
[INFO] |  |  +- com.google.guava:failureaccess:jar:1.0.3:compile
[INFO] |  |  +- com.google.guava:listenablefuture:jar:9999.0-empty-to-avoid-conflict-with-guava:compile
[INFO] |  |  \- com.google.j2objc:j2objc-annotations:jar:3.1:compile
[INFO] |  +- com.google.errorprone:error_prone_annotations:jar:2.45.0:compile
[INFO] |  +- io.perfmark:perfmark-api:jar:0.27.0:runtime
[INFO] |  +- io.grpc:grpc-api:jar:1.79.0:compile
[INFO] |  \- org.codehaus.mojo:animal-sniffer-annotations:jar:1.26:compile
```


Fory: 1.3.0
Java25

### Component(s)

Java

### Minimal reproduce step

add to dependencies io.grpc:grpc-netty-shaded:jar:1.79.0 without any other guava in deps

### What did you expect to see?

Now in `GuavaCollectionSerializers` GUAVA_AVAILABLE calculated for only 6 classes, but in `registerDefaultSerializers` registered a bit more. So it must skip some loading or check each class existance

### What did you see instead?

exception

### Anything Else?


Current workaround for me is exclude guava from grpc-shaded and add dependency for `-jre` instead of `-android` guava
```
        <dependency>
            <groupId>org.apache.fory</groupId>
            <artifactId>fory-core</artifactId>
            <version>1.3.0</version>
        </dependency>

        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>33.5.0-jre</version>
        </dependency>

        <dependency>
            <groupId>io.grpc</groupId>
            <artifactId>grpc-netty-shaded</artifactId>
            <version>${grpc.version}</version>
            <scope>runtime</scope>
            <exclusions>
                <exclusion>
                    <groupId>com.google.guava</groupId>
                    <artifactId>guava</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```


### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

--- a/java/fory-core/src/main/java/org/apache/fory/serializer/collection/GuavaCollectionSerializers.java
+++ b/java/fory-core/src/main/java/org/apache/fory/serializer/collection/GuavaCollectionSerializers.java
@@ -72,6 +72,10 @@ public class GuavaCollectionSerializers {
           && isClassAvailable(IMMUTABLE_SET_CLASS_NAME)
           && isClassAvailable(IMMUTABLE_SORTED_MAP_CLASS_NAME)
           && isClassAvailable(IMMUTABLE_SORTED_SET_CLASS_NAME);
+  private static final boolean SINGLETON_IMMUTABLE_BI_MAP_AVAILABLE =
+      GUAVA_AVAILABLE && isClassAvailable(PKG + ".SingletonImmutableBiMap");
+  private static final boolean SINGLETON_IMMUTABLE_LIST_AVAILABLE =
+      GUAVA_AVAILABLE && isClassAvailable(PKG + ".SingletonImmutableList");
 
   private interface MapEntryBuilder {
     void put(Object key, Object value);
@@ -84,6 +88,11 @@ public class GuavaCollectionSerializers {
       typeResolver.setSerializer(cls, this);
     }
 
+    GuavaCollectionSerializer(TypeResolver typeResolver, Class<T> cls, int ownerBytes) {
+      super(typeResolver, cls, true, ownerBytes);
+      typeResolver.setSerializer(cls, this);
+    }
+
     protected abstract T xnewInstance(Collection collection);
   }
 
@@ -91,6 +100,10 @@ public class GuavaCollectionSerializers {
       extends GuavaCollectionSerializer<T> {
     public ImmutableListSerializer(TypeResolver typeResolver, Class<T> cls) {
       super(typeResolver, cls);
+    }
+
+    ImmutableListSerializer(TypeResolver typeResolver, Class<T> cls, int ownerBytes) {
+      super(typeResolver, cls, ownerBytes);
     }
 
     @Override
@@ -241,6 +254,11 @@ public class GuavaCollectionSerializers {
       typeResolver.setSerializer(cls, this);
     }
 
+    GuavaMapSerializer(TypeResolver typeResolver, Class<T> cls, int ownerBytes) {
+      super(typeResolver, cls, true, ownerBytes);
+      typeResolver.setSerializer(cls, this);
+    }
+
     protected abstract ImmutableMap.Builder makeBuilder(int size);
 
     @Override
@@ -368,6 +386,10 @@ public class GuavaCollectionSerializers {
 
     public ImmutableBiMapSerializer(TypeResolver typeResolver, Class<T> cls) {
       super(typeResolver, cls);
+    }
+
+    ImmutableBiMapSerializer(TypeResolver typeResolver, Class<T> cls, int ownerBytes) {
+      super(typeResolver, cls, ownerBytes);
     }
 
     @Override
@@ -661,14 +683,36 @@ public class GuavaCollectionSerializers {
     Class cls =
         loadClass(PKG + ".RegularImmutableBiMap", ImmutableBiMap.of("k1", 1, "k2", 4).getClass());
     resolver.registerInternalSerializer(cls, new ImmutableBiMapSerializer(resolver, cls));
-    cls = loadClass(PKG + ".SingletonImmutableBiMap", ImmutableBiMap.of(1, 2).getClass());
-    resolver.registerInternalSerializer(cls, new ImmutableBiMapSerializer(resolver, cls));
+    if (SINGLETON_IMMUTABLE_BI_MAP_AVAILABLE) {
+      cls = loadClass(PKG + ".SingletonImmutableBiMap", ImmutableBiMap.of(1, 2).getClass());
+      resolver.registerInternalSerializer(cls, new ImmutableBiMapSerializer(resolver, cls));
+    } else {
+      // This class only preserves the missing singleton ID slot. Charge the concrete Guava owner
+      // returned by the local factory, which may be the regular implementation.
+      class GuavaSingletonBiMap {}
+
+      cls = GuavaSingletonBiMap.class;
+      int ownerBytes = GraphMemoryEstimates.shallowObjectBytes(ImmutableBiMap.of(1, 2).getClass());
+      resolver.registerInternalSerializer(
+          cls, new ImmutableBiMapSerializer(resolver, cls, ownerBytes));
+    }
     cls = loadClass(PKG + ".RegularImmutableMap", ImmutableMap.of("k1", 1, "k2", 2).getClass());
     resolver.registerInternalSerializer(cls, new ImmutableMapSerializer(resolver, cls));
     cls = loadClass(PKG + ".RegularImmutableList", ImmutableList.of().getClass());
     resolver.registerInternalSerializer(cls, new RegularImmutableListSerializer(resolver, cls));
-    cls = loadClass(PKG + ".SingletonImmutableList", ImmutableList.of(1).getClass());
-    resolver.registerInternalSerializer(cls, new ImmutableListSerializer(resolver, cls));
+    if (SINGLETON_IMMUTABLE_LIST_AVAILABLE) {
+      cls = loadClass(PKG + ".SingletonImmutableList", ImmutableList.of(1).getClass());
+      resolver.registerInternalSerializer(cls, new ImmutableListSerializer(resolver, cls));
+    } else {
+      // This class only preserves the missing singleton ID slot. Charge the concrete Guava owner
+      // returned by the local factory, which may be the regular implementation.
+      class GuavaSingletonList {}
+
+      cls = GuavaSingletonList.class;
+      int ownerBytes = GraphMemoryEstimates.shallowObjectBytes(ImmutableList.of(1).getClass());
+      resolver.registerInternalSerializer(
+          cls, new ImmutableListSerializer(resolver, cls, ownerBytes));
+    }
     cls = loadClass(PKG + ".RegularImmutableSet", ImmutableSet.of(1, 2).getClass());
     resolver.registerInternalSerializer(cls, new ImmutableSetSerializer(resolver, cls));
     cls = loadClass(PKG + ".SingletonImmutableSet", ImmutableSet.of(1).getClass());

--- a/java/fory-core/src/test/java/org/apache/fory/serializer/collection/CollectionSerializersTest.java
+++ b/java/fory-core/src/test/java/org/apache/fory/serializer/collection/CollectionSerializersTest.java
@@ -25,6 +25,7 @@ import static org.apache.fory.collection.Collections.ofHashSet;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertSame;
 
+import com.google.common.collect.ImmutableBiMap;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSortedSet;
@@ -77,6 +78,7 @@ import org.apache.fory.config.Int64Encoding;
 import org.apache.fory.config.Language;
 import org.apache.fory.context.ReadContext;
 import org.apache.fory.exception.DeserializationException;
+import org.apache.fory.exception.InsecureException;
 import org.apache.fory.exception.SerializationException;
 import org.apache.fory.memory.MemoryBuffer;
 import org.apache.fory.memory.MemoryUtils;
@@ -85,6 +87,7 @@ import org.apache.fory.platform.JdkVersion;
 import org.apache.fory.reflect.FieldAccessor;
 import org.apache.fory.reflect.TypeRef;
 import org.apache.fory.resolver.TypeResolver;
+import org.apache.fory.serializer.GraphMemoryEstimates;
 import org.apache.fory.serializer.collection.CollectionSerializers.JDKCompatibleCollectionSerializer;
 import org.apache.fory.test.bean.Cyclic;
 import org.apache.fory.type.GenericType;
@@ -1810,6 +1813,65 @@ public class CollectionSerializersTest extends ForyTestBase {
     buffer.readerIndex(0);
     assertEquals(set, fory.deserialize(fory.serialize(buffer, set)));
   }
+
+  @Test
+  public void testGuavaListOwnerBudget() {
+    ImmutableList<String> list = ImmutableList.of("fory");
+    int ownerBytes = GraphMemoryEstimates.shallowObjectBytes(list.getClass());
+    long requiredBytes = ownerBytes + GraphMemoryEstimates.REFERENCE_BYTES;
+    MemoryBuffer buffer = MemoryUtils.buffer(32);
+    Fory fory = graphBudgetFory(requiredBytes);
+    GuavaCollectionSerializers.ImmutableListSerializer serializer =
+        new GuavaCollectionSerializers.ImmutableListSerializer(
+            fory.getTypeResolver(), SingletonListSlot.class, ownerBytes);
+    writeSerializer(fory, serializer, buffer, list);
+
+    Fory limitedFory = graphBudgetFory(requiredBytes - 1);
+    GuavaCollectionSerializers.ImmutableListSerializer limitedSerializer =
+        new GuavaCollectionSerializers.ImmutableListSerializer(
+            limitedFory.getTypeResolver(), SingletonListSlot.class, ownerBytes);
+    Assert.assertThrows(
+        InsecureException.class, () -> readSerializer(limitedFory, limitedSerializer, buffer));
+    buffer.readerIndex(0);
+    assertEquals(readSerializer(fory, serializer, buffer), list);
+  }
+
+  @Test
+  public void testGuavaBiMapOwnerBudget() {
+    ImmutableBiMap<String, Integer> biMap = ImmutableBiMap.of("fory", 1);
+    int ownerBytes = GraphMemoryEstimates.shallowObjectBytes(biMap.getClass());
+    long requiredBytes = ownerBytes + 2L * GraphMemoryEstimates.REFERENCE_BYTES;
+    MemoryBuffer buffer = MemoryUtils.buffer(32);
+    Fory fory = graphBudgetFory(requiredBytes);
+    GuavaCollectionSerializers.ImmutableBiMapSerializer serializer =
+        new GuavaCollectionSerializers.ImmutableBiMapSerializer(
+            fory.getTypeResolver(), SingletonBiMapSlot.class, ownerBytes);
+    writeSerializer(fory, serializer, buffer, biMap);
+
+    Fory limitedFory = graphBudgetFory(requiredBytes - 1);
+    GuavaCollectionSerializers.ImmutableBiMapSerializer limitedSerializer =
+        new GuavaCollectionSerializers.ImmutableBiMapSerializer(
+            limitedFory.getTypeResolver(), SingletonBiMapSlot.class, ownerBytes);
+    Assert.assertThrows(
+        InsecureException.class, () -> readSerializer(limitedFory, limitedSerializer, buffer));
+    buffer.readerIndex(0);
+    assertEquals(readSerializer(fory, serializer, buffer), biMap);
+  }
+
+  private static Fory graphBudgetFory(long maxGraphMemoryBytes) {
+    return Fory.builder()
+        .withXlang(false)
+        .registerGuavaTypes(true)
+        .requireClassRegistration(false)
+        .suppressClassRegistrationWarnings(true)
+        .withCompatible(false)
+        .withMaxGraphMemoryBytes(maxGraphMemoryBytes)
+        .build();
+  }
+
+  private static final class SingletonListSlot {}
+
+  private static final class SingletonBiMapSlot {}
 
   @Test(dataProvider = "foryCopyConfig")
   public void testJavaSerialization(Fory fory) {

